File: 29. Structure pointer demo.py
"""
Structure pointer demo (Python class and list of objects)
"""
class Person:
    def __init__(self, name, age):
        self.name = name; self.age = age

def main():
    p = Person('Alice', 30)
    pp = p
    print(pp.name, pp.age)
    arr = [Person('Bob',25), Person('Carol',28)]
    for person in arr: print(person.name, person.age)

if __name__ == '__main__':
    main()

# Time Complexity: O(n).
# Space Complexity: O(n).

File: 29. Structure pointer demo.cpp
// Topic: Structure pointer: access members using -> operator
#include <iostream>
using namespace std;

struct Person
{
    const char* name;
    int age;
};

int main()
{
    Person p = {"Alice", 30};
    Person* pp = &p;
    cout << pp->name << " " << pp->age << '\n';
    // Array of structure pointers
    Person** arr = new Person*[2];
    arr[0] = new Person{"Bob", 25};
    arr[1] = new Person{"Carol", 28};
    for (int i = 0; i < 2; ++i) cout << arr[i]->name << ' ' << arr[i]->age << '\n';
    for (int i = 0; i < 2; ++i) delete arr[i];
    delete[] arr;
    return 0;
}

// Time Complexity: O(n) for allocation/printing where n is number of structures.
// Space Complexity: O(n) for allocated structures.
